File: Math/CPP/cells_with_odd_values_in_a_matrix.cpp
#include <algorithm>
#include <vector>
using namespace std;

class Solution {
public:
    int oddCells(int m, int n, vector<vector<int>>& indices) {
        vector<int> rows(m, 0);
        vector<int> cols(n, 0);

        for (const auto& idx : indices) {
            rows[idx[0]] ^= 1;
            cols[idx[1]] ^= 1;
        }

        int oddR = count(rows.begin(), rows.end(), 1);
        int oddC = count(cols.begin(), cols.end(), 1);
        int evenR = m - oddR;
        int evenC = n - oddC;
        return oddR * evenC + evenR * oddC;
    }
};

File: Math/Python/cells_with_odd_values_in_a_matrix.py
class Solution:
    def rfill(self, matrix: list, r: int, n):
        for i in range(n):
            matrix[r][i] += 1

    def cfill(self, matrix: list, c: int, m):
        for i in range(m):
            matrix[i][c] += 1

    def oddCells(self, m: int, n: int, indices: list[list[int]]) -> int:
        matrix = [[0] * n for _ in range(m)]
        ans = 0
        for r, c in indices:
            self.rfill(matrix, r, n)
            self.cfill(matrix, c, m)

        for i in range(len(matrix)):
            ans += sum(n & 1 for n in matrix[i])

        return ans


    # simpler one
    def oddCells(self, m: int, n: int, indices: list[list[int]]) -> int:
        # rows[i] and cols[i] :=
        #   1. True (flipped even times)
        #   2. False (flipped odd times)
        rows = [False] * m
        cols = [False] * n

        for r, c in indices:
            rows[r] ^= True
            cols[c] ^= True

        oddRowsCount = rows.count(True)
        oddColsCount = cols.count(True)
        evenRowsCount = m - oddRowsCount
        evenColsCount = n - oddColsCount
        return oddRowsCount * evenColsCount + oddColsCount * evenRowsCount
